subsys/random/rand32_ctr_drbg.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2017, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <device.h>
+#include <entropy.h>
+#include <kernel.h>
+#include <string.h>
+
+#if defined(CONFIG_MBEDTLS)
+#if !defined(CONFIG_MBEDTLS_CFG_FILE)
+#include "mbedtls/config.h"
+#else
+#include CONFIG_MBEDTLS_CFG_FILE
+#endif /* CONFIG_MBEDTLS_CFG_FILE */
+#include <mbedtls/ctr_drbg.h>
+
+#elif defined(CONFIG_TINYCRYPT)
+
+#include <tinycrypt/ctr_prng.h>
+#include <tinycrypt/aes.h>
+#include <tinycrypt/constants.h>
+
+#endif /* CONFIG_MBEDTLS */
+
+static K_SEM_DEFINE(state_sem, 1, 1);
+
+static struct device *entropy_driver;
+
+#if defined(CONFIG_MBEDTLS)
+
+static mbedtls_ctr_drbg_context ctr_ctx;
+
+static int ctr_drbg_entropy_func(void *ctx, unsigned char *buf, size_t len)
+{
+	return entropy_get_entropy(ctx, (void *)buf, len);
+}
+
+#elif defined(CONFIG_TINYCRYPT)
+
+static TCCtrPrng_t ctr_ctx;
+
+#endif /* CONFIG_MBEDTLS */
+
+
+static int ctr_drbg_initialize(void)
+{
+	int ret;
+	static const unsigned char drbg_seed[] = "zephyr ctr-drbg seed";
+
+	/* Only one entropy device exists, so this is safe even
+	 * if the whole operation isn't atomic.
+	 */
+	entropy_driver = device_get_binding(CONFIG_ENTROPY_NAME);
+	if (!entropy_driver) {
+		__ASSERT((entropy_driver != NULL),
+			"Device driver for %s (CONFIG_ENTROPY_NAME) not found. "
+			"Check your build configuration!",
+			CONFIG_ENTROPY_NAME);
+		return -EINVAL;
+	}
+
+#if defined(CONFIG_MBEDTLS)
+
+	mbedtls_ctr_drbg_init(&ctr_ctx);
+
+	ret = mbedtls_ctr_drbg_seed(&ctr_ctx, ctr_drbg_entropy_func,
+				    entropy_driver,
+				    drbg_seed, sizeof(drbg_seed));
+
+	if (ret != 0) {
+		mbedtls_ctr_drbg_free(&ctr_ctx);
+		return -EIO;
+	}
+
+#elif defined(CONFIG_TINYCRYPT)
+
+	u8_t entropy[TC_AES_KEY_SIZE + TC_AES_BLOCK_SIZE];
+
+	entropy_get_entropy(entropy_driver, (void *)&entropy, sizeof(entropy));
+
+	ret = tc_ctr_prng_init(&ctr_ctx,
+			       (uint8_t *)&entropy,
+			       sizeof(entropy),
+			       (uint8_t *)drbg_seed,
+			       sizeof(drbg_seed));
+
+	if (ret == TC_CRYPTO_FAIL) {
+		return -EIO;
+	}
+
+#endif
+
+	k_object_access_all_grant(&state_sem);
+
+	return 0;
+}
+
+
+u32_t sys_csrand32_get(void)
+{
+	u32_t ret;
+
+	if (unlikely(!entropy_driver)) {
+		ctr_drbg_initialize();
+	}
+
+
+	if (k_sem_take(&state_sem, K_NO_WAIT) < 0) {
+		/*
+		 * If the CSPRNG is blocked then try the non-CSPRNG.
```

Comment:
Um, no. Aside from that nobody should be calling this function, you don't ***ever*** want a situation where the csprng can be tricked into returning non-cryptographic random data. There are fairly easy attacks that would cause the device to be busy and this to always or frequently fall through to the weak path.